File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.js
const input = require('fs').readFileSync('/dev/stdin').toString().split('\n');
const [N,K] = input[0].split(' ').map(Number);
let dp = new Array(K+1).fill(0); // 인덱스:무게 의미 & 값:가치 합산

// 반복해서 입력데이터들에 대해 우선 데이터 갱신
for(let i=1; i<=N; i++) {
    const [W,V] = input[i].split(' ').map(Number);
    
    // 뒤에서부터 반복해 현재 물건의 무게를 고려하여 최대 가치 계산
    for (let j = K; j >= W; j--) {
        dp[j] = Math.max(dp[j], dp[j - W] + V);
    }
}

// 출력: K인덱스에 담긴 값 출력
console.log(dp[K]);

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline

N, K = map(int, input().strip().split())
items = [tuple(map(int, input().strip().split())) for _ in range(N)]
dp = [[-1] * (K + 1) for _ in range(N + 1)]

def solve(i, w, dp):
    if i == 0 or w == 0:
        return 0
    
    if dp[i][w] != -1:
        return dp[i][w]
    
    weight, value = items[i - 1]
    if weight > w:
        dp[i][w] = solve(i - 1, w, dp)
    else:
        dp[i][w] = max(solve(i - 1, w, dp), solve(i - 1, w - weight, dp) + value)
        
    return dp[i][w]
    
# 결과 출력
print(solve(N, K, dp))